### PR TITLE
[CPU] Break LLVMCPUVectorLowering pass to several small passes.

### PR DESCRIPTION
**Pull request by hanhanW**
The pass is broken into four passes:

- LLVMCPUVirtualVectorLowering
- LLVMCPUVectorTransferLowering
- LLVMCPUVectorTransposeLowering
- LLVMCPUVectorShapeCastLowering

The revision creates a vector lowering pipeline, and addresses TODO about running canonicalizer.

The original big pass had four stage. There were four applyPatternsAndFoldGreedily, and they stop parallelism. The IR verification is what we want. Because it makes sure that invalid IR is not generated between each stage. We also need a full canonicalizer between two of stages, and it serves the need.

It is a step towards (3) in https://github.com/openxla/iree/issues/16025#issuecomment-1881908210

benchmark-extra: x86_64-dt-only

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUVectorLowering.cpp [hunk 1]
```diff
@@ -1,195 +0,0 @@
...
-  LLVM_DEBUG({
-    llvm::dbgs() << "\n--- After lowering high level vector ops to lower level "
-                    "vector ops ---\n";
-    funcOp.print(llvm::dbgs(), OpPrintingFlags().useLocalScope());
-    llvm::dbgs() << "\n\n";
-  });
-
-  // Make sure we remove redundant vector ops (e.g., vector tranposes) before we
-  // lower them and can't be optimized away anymore.
-  // TODO (dcaballe): We should run full canonicalization here.
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUVectorLowering.cpp:123]: This is the TODO item that addressed by the PR

**Comment by benvanik** [comment]
driveby note that each pass adds to the compilation time (the verifier runs between each pass, passes don't run in parallel, etc) - would be good to balance code health with that to make sure we don't end up with 1000 passes that each apply a pattern or two :)

**Comment by hanhanW** [comment]
@dcaballe I take a stab at breaking this big pass, please take a look. I'm pretty sure that there should be some follow ups (e.g., I added a TODO in VirtualVectorLowering), but this is a good step to me. Please also advice if you have better naming about the new passes. :)

**Comment by hanhanW** [comment]
> driveby note that each pass adds to the compilation time (the verifier runs between each pass, passes don't run in parallel, etc) - would be good to balance code health with that to make sure we don't end up with 1000 passes that each apply a pattern or two :)

Thanks! I will keep it in mind. Here are more notes about the change:

The original big pass had four stage. There were four `applyPatternsAndFoldGreedily`, and they stop parallelism. The IR verification is what we want. Because it makes sure that invalid IR is not generated between each stage. We also need a full canonicalizer between two of stages, and it serves the need.

IMO, vector lowering is a giant monster and it is hard. We want to make sure that we are progressively lowering vectors; we don't want to mix all the patterns together. We have four passes today. We could end up with 4-6 passes, but not something like 1000 passes. :-)

File: compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUVectorTransposeLowering.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUVectorTransposeLowering.cpp:1]: @Max191 heads-up that the bitcast patterns for transpose lowering that we've discussed should be added to this file.

**Comment by MaheshRavishankar** [comment]
> > driveby note that each pass adds to the compilation time (the verifier runs between each pass, passes don't run in parallel, etc) - would be good to balance code health with that to make sure we don't end up with 1000 passes that each apply a pattern or two :)
> 
> Thanks! I will keep it in mind. Here are more notes about the change:
> 
> The original big pass had four stage. There were four `applyPatternsAndFoldGreedily`, and they stop parallelism. The IR verification is what we want. Because it makes sure that invalid IR is not generated between each stage. We also need a full canonicalizer between two of stages, and it serves the need.
> 
> IMO, vector lowering is a giant monster and it is hard. We want to make sure that we are progressively lowering vectors; we don't want to mix all the patterns together. We have four passes today. We could end up with 4-6 passes, but not something like 1000 passes. :-)

Yeah +1 to this. We've always wanted to clean up the vectorization flow better instead of having a kitchen sink of throw everything in approach. Compile time concerns are valid, but I would rather pay that here for getting a better handle on vectorization.

**Review by MaheshRavishankar — APPROVED** [review]

**Comment by benvanik** [comment]
sounds worth it to me - I'm not familiar with this code and just noticed a bunch of new passes applying a single set of patterns - sounds like it's better than before, but I wouldn't want to see this kind of approach spread to other places as it indicates larger issues

**Comment by hanhanW** [comment]
> I wouldn't want to see this kind of approach spread to other places as it indicates larger issues

AFAIK, this should only happen on vector lowering atm. We could improve it by defining more proper dialects, but not sure how to do this. I don't want to see it in other places either. :)

**Pull request merged**